docs/proposals/templates.md

```diff
@@ -0,0 +1,400 @@
+
+#### Template Authoring
+
+To aid application authors in the creation of new templates, it should be possible to export existing objects from a project 
+in template form.  A user should be able to export all or a filtered subset of objects from a namespace, wrappered into a 
+Template API object.  The user will still need to customize the resulting object to enable parameterization and labeling, 
+though sophisticated export logic could attempt to auto-parameterize well understood api fields.  Such logic is not considered 
+in this proposal.
+
+#### Tooling
+As described above, templates can be instantiated by posting them to a template processing endpoint.  CLI tools should 
+exist which can input parameter values from the user as part of the template instantiation flow.  
+
+More sophisticated UI implementations should also guide the user through which parameters the template expects, the description 
+of those templates, and the collection of user provided values.
+
+In addition, as described above, existing objects in a namespace can be exported in template form, making it easy to recreate a 
+set of objects in a new namespace or a new cluster.
+
+
+## Examples
+
+#### Example Templates
+
+* [Jenkins template](https://github.com/openshift/origin/blob/master/examples/jenkins/jenkins-persistent-template.json)
+* [MySQL DB service template](https://github.com/openshift/origin/blob/master/examples/db-templates/mysql-persistent-template.json)
+
+#### Examples of OpenShift Parameter Usage
+
+(mapped to use cases described above)
+
+* [Generate and share passwords](https://github.com/jboss-openshift/application-templates/blob/master/eap/eap64-mongodb-s2i.json#L146-L152)
+* [Simple deployment-time customization of “app” configuration via environment values](https://github.com/jboss-openshift/application-templates/blob/master/eap/eap64-mongodb-s2i.json#L108-L126) (eg memory tuning, resource limits, etc)
+* [Customization of component names with referential integrity](https://github.com/jboss-openshift/application-templates/blob/master/eap/eap64-mongodb-s2i.json#L199-L207)
+* [Customize cross-component references](https://github.com/jboss-openshift/application-templates/blob/master/eap/eap64-mongodb-s2i.json#L78-L83) (eg user provides the name of a secret that already exists in their namespace, to use in a pod as a TLS cert)
+
+## Known Limitations
+
+* Generators must be statically compiled into the template api endpoint
+* Currently only string fields can be parameterized.  See discussion [here](https://github.com/openshift/origin/issues/3946) for some proposed
+  solutions. (tldr: a syntax like ${FOO | int} could instruct the template processor to convert the json field to an integer when doing the
+  substitution.  This could also support secret generation by using a syntax like `${FOO | base64}` which would take the value of parameter 
+  `$FOO` and base64 encode it before setting the json field to the resulting value)
```

Comment:
Definitely need a solution for this.  Functor pipelines start to increase the complexity of the proposal and the line between template and language gets blurry.